### PR TITLE
pkp/pkp-lib#9434 Allow plugin level API call

### PR DESCRIPTION
**Pull request by touhidurabir**
for pkp/pkp-lib#9434

### CODE REVIEW

File: classes/core/APIRouter.php [hunk 1]
```diff
@@ -3,8 +3,8 @@
 /**
  * @file classes/core/APIRouter.php
  *
- * Copyright (c) 2014-2021 Simon Fraser University
- * Copyright (c) 2000-2021 John Willinsky
+ * Copyright (c) 2023 Simon Fraser University
+ * Copyright (c) 2023 John Willinsky
```
Comment (asmecher) [inline classes/core/APIRouter.php:7]: (I'm not sure it's at all necessary, but our general practice is to extend the date when editing classes -- so these should be changed to `2014-2023` and `2000-2023`, not just `2023`.)

**Review by asmecher — CHANGES_REQUESTED** [review]
This is not a bad idea, @touhidurabir, but can we come back to it later? We don't yet have a required use case for a plugin to add its own namespace to the API, and this does build in greater complexity to the `APIHandler`. At some point (I hope sooner than later) we can continue to replace our own routing/dispatching code with Laravel's, building on your work with the Laravel API router; when we do that, the fewer cases we have to replicate in Laravel, the better.

**Comment by touhidurabir** [comment]
> This is not a bad idea, @touhidurabir, but can we come back to it later? We don't yet have a required use case for a plugin to add its own namespace to the API, and this does build in greater complexity to the APIHandler.

@asmecher I agree . This is something I did as just to show a proof of concept that allow us now an easier way to have this functionality . 

BTW, in the previous implementation of 3.3.0/3.4.0, we have hook to allow plugin to intervene the API endpoints , see at https://github.com/pkp/pkp-lib/blob/stable-3_4_0/classes/handler/APIHandler.php#L192 . Now that we have moved to `Laravel Routing System` it's been removed and previous approach will not work . Do we need such functionality ? If so, we can bring in that part only which is included in this PR(in that case I can create a new issue and only port the changes for the part) . If no needed, we can leave it for now .

File: classes/core/APIRouter.php [hunk 2]
```diff
@@ -90,14 +187,22 @@ public function route(PKPRequest $request): void
...
         $this->setHandler($handler);
+        $handler->runRoutes();
     }
 
     /**
      * Get the requested operation
+     *
+     * @return string
      */
-    public function getRequestedOp(PKPRequest $request): string
```
Comment (asmecher) [inline classes/core/APIRouter.php:100]: Is this an accidental revert of adding typehints?

File: classes/core/PKPBaseController.php [hunk 3]
```diff
@@ -114,8 +113,19 @@ public static function getRouteController(Request $request = null): ?static
         if (!$requestedRoute = static::getRequestedRoute($request)) {
             return null;
         }
+        
+        $calledRouteController = (new ReflectionFunction($requestedRoute->action['uses']))->getClosureThis();
+
+        // When the routes are added to router as a closure/callbale from other section like from a
```
Comment (asmecher) [inline classes/core/PKPBaseController.php:119]: *callable

File: classes/handler/APIHandler.php [hunk 4]
```diff
@@ -105,23 +144,66 @@ public function __construct(PKPBaseController $controller)
...
+            'roles' => $roles
+        ]);
+    }
+
+    /**
+     * Register the routes in the routes collection which was added via hook
+     */
+    protected function registerRoute(): void
+    {
+        if (empty($this->routesFromHook)) {
```
Comment (asmecher) [inline classes/handler/APIHandler.php:190]: I think these 3 lines can be removed.

**Review by asmecher — CHANGES_REQUESTED** [review]
Just a few minor comments -- thanks, @touhidurabir!

File: classes/core/APIRouter.php [hunk 5]
```diff
@@ -156,10 +179,8 @@ public function url(
         //
         // Assemble URL
         //
-        $pathInfoArray = array_merge(
-            $context,
-            ['api', Application::API_VERSION, $endpoint]
-        );
+        $pathInfoArray = ['api', Application::API_VERSION, $endpoint];
+        $pathInfoArray = array_merge($context, $pathInfoArray);
```
[the remark this answers is not in the document]
Comment (touhidurabir) [reply classes/core/APIRouter.php]: rebased and reverted back .

File: classes/core/APIRouter.php [hunk 6]
```diff
@@ -23,31 +21,47 @@
...
+        if (is_null($pathInfoParts)) {
+            return false;
+        }
+
+        if (count($pathInfoParts) < 2) {
+            return false;
+        }
+
+        // Context-specific API requests: [index.php]/{contextPath}/api
+        if (strtolower($pathInfoParts[1]) === 'api') {
```
[the remark this answers is not in the document]
Comment (touhidurabir) [reply classes/core/APIRouter.php]: good catch , it's a leftover change from removing plugin own level api

File: classes/core/Dispatcher.php [hunk 7]
```diff
@@ -198,15 +196,16 @@ public function setUserResolver(): void
...
-     * @param $shortcut the short name of the router that should be used to construct the URL
-     * @param $newContext Optional contextual path
-     * @param $handler Optional name of the handler to invoke
-     * @param $op Optional name of operation to invoke
-     * @param $path Optional string or array of args to pass to handler
-     * @param $params Optional set of name => value pairs to pass as user parameters
-     * @param $anchor Optional name of anchor to add to URL
-     * @param $escape Whether or not to escape ampersands for this URL; default false.
-     * @param $urlLocaleForPage Whether or not to override locale for this URL; Use '' to exclude.
+     * @param PKPRequest    $request            The request object
```
[the remark this answers is not in the document]
Comment (touhidurabir) [reply classes/core/Dispatcher.php]: reverted back

File: classes/handler/APIHandler.php (hunk 4 above)
Comment (touhidurabir) [reply classes/handler/APIHandler.php]: removed .

**Comment by touhidurabir** [comment]
@asmecher all requested changes have been addressed. I see that you have already added the removed hook `APIHandler::endpoints` in the `Hook::addUnsupportedHooks` .

**Comment by asmecher** [comment]
Thanks, @touhidurabir; this needs a rebase before it can be merged, and could you also run it past the developer who was looking for the feature, along with a link to your example plugin?

**Pull request merged**